src/OrchardCore.Modules/OrchardCore.GitHub/Configuration/GithubHandler.cs

```diff
@@ -36,5 +44,65 @@ protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIden
             await Events.CreatingTicket(context);
             return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
         }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(OAuthCodeExchangeContext context)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", context.RedirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", context.Code },
+                { "grant_type", "authorization_code" },
+            };
+
+            // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+            if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+            {
+                tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            requestMessage.Content = requestContent;
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode)
+            {
+                var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+
+
+                if(payload.RootElement.TryGetProperty("error", out var error))
+                {
+                    var output = new StringBuilder();
+                    output.Append(error);
+
+                    if(payload.RootElement.TryGetProperty("error_description", out var description))
+                    {
+                        output.Append(" ");
+                        output.Append(description);
+                    }
+                    return OAuthTokenResponse.Failed(new Exception(output.ToString()));
```

Comment:
It's quite a lot of dependant code to copy in, to get the custom error message. 

Did you consider a delegating http handler on the `Options.BackChannel` httpclient ?

Or perhaps it is worth opening an issue on ASP.NET to see about the possiblities of better error handling overrides?

Reply:
I agree with you, too much dependant code to copy. 

I think ASP.NET should have an extensibility point to parse and validate the OAuth response.